[PT2/Profiler] Add Context Info to Torch-Compiled Regions

sraikund16 opened the pull request:
Summary: 
We want to add compile IDs and frames to each Torch-Compiled Region in order to help users cross reference the section they are checking alongside data obtained from tools, such as tlparse.
This diff operates on the assumption that each graph section will enter and exit a CompileContext before it is ran to either compile the graph or look it up in the cache. Based on this assuption, we can save the value of the graph section from the exited CompileContext in eval_frame.c using a Python C API. After this, we can create a new interface in cpp shim to wrap around the record_function in order to pass in the new keyword argument for "context".

Test Plan: 
Enhance test_profiler_dynamo_compiled_region to look for kwinputs as well as a name to see that the context is now labeled. Also changed test to run graph with more contexts so that we test a wider range of profiling.

Differential Revision: D60803317


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

davidberard98 commented:
cc @williamwen42 @yanboliang for the dynamo parts

also, if you can get a local dynamo benchmark setup running, if you can test the impact this has on the `BERT_pytorch` model, that would be great! or if not, you can try with the performance benchmark dashboard (https://hud.pytorch.org/benchmark/compilers). This would be a good quick test to see if there's any perf impact from this change.

williamwen42 commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -476,14 +476,36 @@ inline static PyObject* eval_custom_code_impl(
     int throw_flag,
     int free_vars_copied) {
-  _PytorchRecordFunctionState* rf = _pytorch_record_function_enter("Torch-Compiled Region");
+  const char* trace_id = get_compile_context(tstate);
```
This is a hot function. Can we avoid calling this when unnecessary and/or cache the result?

sraikund16 replied:
I think if we store a global as you mentioned in the comment on line 479 it will fix this bottleneck

williamwen42 commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -476,14 +476,36 @@ inline static PyObject* eval_custom_code_impl(
 
+static const char* get_compile_context(PyThreadState* tstate) {
+    // Import the torch module
+    PyObject* torch_module = PyImport_ImportModule("torch._guards");
```
Prefer having a global in C and defining builtin methods for accessing in Python, rather than having C accessing Python.

sraikund16 replied:
That sounds like a reasonable solution as well. Some groups have stated that they would prefer that we don't cache graph values for use of profiling, as it forces us to update parameters in two locations, between python and C. However, since this affects performance, it sounds like it is a better route to take.

williamwen42 commented on torch/csrc/dynamo/cpp_shim.cpp:
```diff
@@ -14,6 +13,24 @@ _PytorchRecordFunctionState* _pytorch_record_function_enter(const char* name) {
+_PytorchRecordFunctionState* _pytorch_record_function_enter_with_context(
+    const char* name,
+    const char* context) {
+  auto map = std::make_unique<std::unordered_map<std::string, c10::IValue>>();
```
Does this need to be a unique_ptr? Can we not just make it a stack-allocated unordered_map?

sraikund16 replied:
updated

williamwen42 reviewed (commented):
Could you also make sure that your changes don't significantly increase overhead by running and comparing microbenchmarks (`benchmarks/dynamo/microbenchmarks/overheads.py`)?

sraikund16 commented:
> Could you also make sure that your changes don't significantly increase overhead by running and comparing microbenchmarks (`benchmarks/dynamo/microbenchmarks/overheads.py`)?

@williamwen42 

**Origin Main**
requires_grad=False
eager    7.6us (warmup=0.0s)
compiled 27.1us (warmup=22.8s)

requires_grad=True
eager    9.6us (warmup=0.0s)
compiled 68.3us (warmup=0.2s)

inference_mode()
eager    7.0us (warmup=0.0s)
compiled 20.9us (warmup=0.1s)


**This PR**
requires_grad=False
eager    7.3us (warmup=0.0s)
compiled 27.2us (warmup=22.2s)

requires_grad=True
eager    9.0us (warmup=0.0s)
compiled 65.5us (warmup=0.2s)

inference_mode()
eager    6.9us (warmup=0.0s)
compiled 20.7us (warmup=0.1s)

sraikund16 committed:
[PT2/Profiler] Add Context Info to Torch-Compiled Regions (#132765)

Summary:
We want to add compile IDs and frames to each Torch-Compiled Region in order to help users cross reference the section they are checking alongside data obtained from tools, such as tlparse.

This diff operates on the assumption that each graph section will enter and exit a CompileContext before it is ran to either compile the graph or look it up in the cache. Based on this assuption, we can save the value of the graph section from the exited CompileContext in eval_frame.c using a Python C API. After this, we can create a new interface in cpp shim to wrap around the record_function in order to pass in the new keyword argument for "context".

Pull Request resolved: https://github.com/pytorch/pytorch/pull/132765

Test Plan: Enhance test_profiler_dynamo_compiled_region to look for kwinputs as well as a name to see that the context is now labeled. Also changed test to run graph with more contexts so that we test a wider range of profiling.

Reviewed By: anijain2305

Differential Revision: D60803317

bdhirsh commented on torch/_guards.py:
```diff
@@ -781,6 +782,15 @@ def compile_context(context: Optional[CompileContext]):
     finally:
+        if context is not None:
+            if context.compile_id is not None:
+                set_context_frame(
```
(btw this feature is awesome 😄 )

side car comment, maybe cc @anijain2305 - won't this function get called (and set the context) during compilation? It seems like we would want this state to be set to the "previously compiled frame" **every time** a given compiled object is invoked at runtime.

bdhirsh replied:
hmm yeah, repro here:

```
@torch.compile(backend="aot_eager", dynamic=True)
def f(x):
    if x.shape[0] > 5:
        return x.sin()
    else:
        return x.cos()

x1 = torch.randn(4)
x2 = torch.randn(6)
x3 = torch.randn(3)

with torch.profiler.profile(record_shapes=True) as prof:
    out1 = f(x1)
    out2 = f(x2)
    out3 = f(x3)

print([e.kwinputs['context'] for e in prof.events() if 'Compiled' in e.name])
```

this should print:
```
['0/0', '0/1', '0/0']
```

since we are re-running the first compiled object on the third case. But instead, I get:
```
['0/0', '0/1', '0/1']
```

I can file an issue

bdhirsh replied:
https://github.com/pytorch/pytorch/issues/136235

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -10,9 +10,11 @@
 PyObject* guard_error_hook = NULL;
 const char* cache_lookup_profiler_str = "TorchDynamo Cache Lookup";
-
+static char compile_context[MAX_COMPILE_CONTEXT_SIZE];
```
We should NOT be maintaining global state like this. Superficially, it is not thread safe. But more conceptually, compile id of a compiled object is associated with the compile object itself, it's not something that should be set at runtime. When you compile some code, the compile product is what stores the compile id (0/0, whatever), and you should be pulling out the compile id *from the compile product* itself.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -817,12 +820,27 @@ static PyObject* set_guard_error_hook(PyObject* dummy, PyObject* obj) {
     {"unsupported", unsupported, METH_VARARGS, NULL},
     {"skip_code", skip_code, METH_O, NULL},
     {"set_guard_error_hook", set_guard_error_hook, METH_O, NULL},
+    {"set_context_frame", set_context_frame, METH_O, NULL},
```
When you get rid of global state, you can delete this function entirely. Instead...

ezyang commented on aten/src/ATen/record_function.h:
```diff
@@ -319,7 +319,7 @@ struct TORCH_API RecordFunction {
       return;
     }
-    kwinputs_ = *kwargs;
+    kwinputs_ = std::unordered_map<std::string, IValue>(*kwargs);
```
What's going on here?

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -483,7 +485,8 @@ inline static PyObject* eval_custom_code(
     PyCodeObject* code,
     int throw_flag,
     int free_vars_copied) {
-  _PytorchRecordFunctionState* rf = _pytorch_record_function_enter("Torch-Compiled Region");
```
What I want here, is I want a `const char*` passed in with the name of the compile region to this function, and then we use the old `_pytorch_record_function_enter`. Let me check the call sites to ensure you can do this...

ezyang commented:
Concretely, I think we should revert this PR redo it under a better design that doesn't have the problem @bdhirsh pointed out

ezyang commented:
@pytorchbot revert -c weird -m "implementation is not correct, needs full rewrite"
